cip/1.accepted/CIP2021-07-07-Grouping-keys-and-aggregation-expressions.adoc

```diff
@@ -0,0 +1,779 @@
+Since these expressions are effectively evaluate after all _POST_PROJ_(_p~i~_) expressions, a similar syntax restrictions applies to the <SortItem>s.
+
+However, <SortItem>s can refer to aliases introduced by _POST_PROJ_(_p~i~_).
+
+Given a list of <ProjectionItem>s _L_ = {_p_~1~, _p_~2~, ..., _p~N~_}, let _GROUPING_KEYS_AND_ALIASES_(_L_) be the set of
+
+* all expressions _ex_ for which there is a <ProjectionItem>s _p_ = `_ex_ AS _al_` in _L_ where _AGG_(_p_) is empty and
+* all aliases _al_ for which there is a <ProjectionItem>s _p_ = `_ex_ AS _al_` in _L_.
+
+For an expression _ex_ and a projection list _L_, let _AVAILABLE_TO_ORDER_(_ex_, _L_) hold
+
+* If _ex_ is either
+** A grouping key or an alias, i.e. _ex_ ∈ _GROUPING_KEYS_AND_ALIASES_(_L_),
+** A constant, or
+** A parameter,
+* or if _ex_ comprises of sub-expressions, it only comprises sub-expressions _subEx_ for which _AVAILABLE_TO_ORDER_(_subEx_, _L_) holds.
+
+[IMPORTANT]
+====
+For `WITH _L_ ORDER BY _SI_` and `RETURN _L_ ORDER BY _SI_` and every _ex_ contained in a <SortItem> in _SI_, _AVAILABLE_TO_ORDER_(_ex_, _L_) shall hold.
+====
+
+==== Omitted aliases
+
+This proposal considers all <ProjectionItem>s have user-given alias.
+Cypher allows to omit the aliases, particularly in the RETURN clause, though.
+However, the alias omission rules are based on the assumption that an implementation will infer a more or less reasonable alias if an alias is omitted by the user.
+Hence, it is safe for this proposal to assume that all <ProjectionItem>s have an alias.
+
+=== Examples
+
+==== Valid aggregations
+
+The following clauses exhibit valid aggregations under the grouping and aggregation semantics and the syntax restriction it includes.
+For each example we list why it is valid.
+
+. `RETURN 1 + count(*)`
+* The sub-expression `1` is a constant.
+
+. `RETURN 1, 1 + count(*)`
+* The sub-expression `1` is a constant.
+
+. `RETURN $x + count($x)`
+* The sub-expression `$x` is a parameter.
+
+. `RETURN count($x) + $x`
+* The sub-expression `$x` is a parameter.
+
+. `RETURN 1 + count($x) + $x * 7 + sum($x) + 'cake'`
+* The sub-expressions `1`, `2`, and `'cake'` are constants.
+* The sub-expression `$x` is a parameter.
+
+. `MATCH (a) RETURN a.x, 1 + count(a.x)`
+* The sub-expression `1` is a constant.
+
+. `MATCH (a) RETURN a.x, a.x + count(a.x)`
+* The sub-expression `a.x` is a grouping key.
+
+. `MATCH (a) WITH a.x AS ax RETURN ax, ax + count(ax)`
+* The sub-expression `ax` is a grouping key.
+
+. `MATCH (a)-[]->(b) RETURN a, a.x + count(b.y)`
+* The sub-expression `a` is a grouping key.
+
+. `MATCH (a)-[]->(b) RETURN a, size(keys(a)) + count(b.y)`
+* The sub-expression `a` is a grouping key.
+
+. `MATCH (x) RETURN x.a, x.b, x.c, x.a + x.b + count(x) + x.c`
+* The sub-expressions `x.a`, `x.b`, and `x.c` are grouping keys.
+
+. `MATCH (a) RETURN a.x + 1, a.x + 1 + count(a.x)`
```

Comment:
Should this really be valid? Didn't we say it would be too confusing since we couldn't allow `MATCH (a) RETURN a.x + 1, count(a.x) + a.x + 1 `?

Reply:
That is related to what is written in the caveats section. From the openCypher grammar standpoint there is no left-deep understanding to chained numeric additions. So there is no hook to hang such a restriction on.

I do not like this either, but I think it goes beyond this CIP to change the grammar in that regard. We could escape the problem by saying in the CIP that the CIP assumes a left-deep parse tree. However, the CLG notes did not indicate that this assumption was part of the decision, hence, I did not felt mandated to do it. 

I rather have a CLG discussion on that issues -- that is why I wrote the caveat part.